lib/r10k/module_loader/puppetfile.rb

```diff
@@ -0,0 +1,190 @@
+                     puppetfile: File.join(basedir, DEFAULT_PUPPETFILE_NAME),
+                     forge: DEFAULT_FORGE_API,
+                     overrides: {},
+                     environment: nil)
+
+        @basedir     = basedir
+        @moduledir   = moduledir
+        @puppetfile  = puppetfile
+        @forge       = forge
+        @overrides   = overrides
+        @environment = environment
+
+        @modules = []
+
+        @managed_directories = []
+        @desired_contents = []
+        @purge_exclusions = []
+      end
+
+      def load!
+        dsl = R10K::ModuleLoader::Puppetfile::DSL.new(self)
+        dsl.instance_eval(puppetfile_content(@puppetfile), @puppetfile)
+
+        validate_no_duplicate_names!(@modules)
+        @modules.freeze
+
+        managed_content = @modules.group_by(&:dirname).freeze
+
+        @managed_directories = determine_managed_directories(managed_content).freeze
+        @desired_contents = determine_desired_contents(managed_content).freeze
+        @purge_exclusions = determine_purge_exclusions(@managed_directories.clone).freeze
+
+        {
+          modules: @modules,
+          managed_directories: @managed_directories,
+          desired_contents: @desired_contents,
+          purge_exclusions: @purge_exclusions
+        }.freeze
+
+      rescue SyntaxError, LoadError, ArgumentError, NameError => e
+        raise R10K::Error.wrap(e, _("Failed to evaluate %{path}") % {path: @puppetfile})
+      end
+
+
+      ##
+      ## set_forge, set_moduledir, and add_module are used directly by the DSL class
+      ##
+
+      # @param [String] forge
+      def set_forge(forge)
+        @forge = forge
+      end
+
+      # @param [String] moduledir
+      def set_moduledir(moduledir)
+        @moduledir = if Pathname.new(moduledir).absolute?
+          moduledir
+        else
+          File.join(@basedir, moduledir)
+        end
+      end
+
+      # @param [String] name
+      # @param [Hash, String, Symbol] args Calling with anything but a Hash is
+      #   deprecated. The DSL will now convert String and Symbol versions to
+      #   Hashes of the shape
+      #     { version: <String or Symbol> }
+      #
+      def add_module(name, args)
+        if !args.is_a?(Hash)
+          args = { version: args }
+        end
+
+        args[:overrides] = @overrides
+
+        if install_path = args.delete(:install_path)
+          install_path = resolve_install_path(install_path)
+          validate_install_path!(install_path, name)
+        else
+          install_path = @moduledir
+        end
+
+        if @default_branch_override != nil
+          args[:default_branch_override] = @default_branch_override
+        end
+
+        mod = R10K::Module.new(name, install_path, args, @environment)
+        mod.origin = :puppetfile
+
+        # Do not load modules if they would conflict with the attached
+        # environment
+        if @environment && @environment.module_conflicts?(mod)
```

Comment:
Should we be warning at all in this case?

Reply:
Looking into it more it seems that the `module_conflicts?` method handles that already https://github.com/puppetlabs/r10k/blob/main/lib/r10k/environment/with_modules.rb#L52

Reply:
ahh nice, thanks for finding that.